server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,190 @@
+     */
+    public OnHeapStarTreeBuilder(
+        StarTreeField starTreeField,
+        Map<String, DocValuesProducer> fieldProducerMap,
+        SegmentWriteState segmentWriteState,
+        MapperService mapperService
+    ) throws IOException {
+        super(starTreeField, fieldProducerMap, segmentWriteState, mapperService);
+    }
+
+    @Override
+    public void appendStarTreeDocument(StarTreeDocument starTreeDocument) throws IOException {
+        starTreeDocuments.add(starTreeDocument);
+    }
+
+    @Override
+    public StarTreeDocument getStarTreeDocument(int docId) throws IOException {
+        return starTreeDocuments.get(docId);
+    }
+
+    @Override
+    public List<StarTreeDocument> getStarTreeDocuments() {
+        return starTreeDocuments;
+    }
+
+    @Override
+    public Long getDimensionValue(int docId, int dimensionId) throws IOException {
+        return starTreeDocuments.get(docId).dimensions[dimensionId];
+    }
+
+    @Override
+    public Iterator<StarTreeDocument> sortAndAggregateStarTreeDocuments(int numDocs) throws IOException {
+        StarTreeDocument[] starTreeDocuments = new StarTreeDocument[numDocs];
+        for (int currentDocId = 0; currentDocId < numDocs; currentDocId++) {
+            starTreeDocuments[currentDocId] = getSegmentStarTreeDocument(currentDocId);
+        }
+        return sortAndAggregateStarTreeDocuments(starTreeDocuments);
+    }
+
+    /**
+     * Sort, aggregates and merges the star-tree documents
+     * @param starTreeDocuments star-tree documents
+     * @return iterator for star-tree documents
+     * @throws IOException throws when unable to sort, merge and aggregate star-tree documents
+     */
+    Iterator<StarTreeDocument> sortAndAggregateStarTreeDocuments(StarTreeDocument[] starTreeDocuments) throws IOException {
+
+        // sort the documents
+        Arrays.sort(starTreeDocuments, (o1, o2) -> Arrays.compare(o1.dimensions, o2.dimensions));
+
+        // merge the documents
+        return mergeStarTreeDocuments(starTreeDocuments);
+    }
+
+    /**
+     * Merges the star-tree documents
+     * @param starTreeDocuments star-tree documents
+     * @return iterator to aggregate star-tree documents
+     */
+    private Iterator<StarTreeDocument> mergeStarTreeDocuments(StarTreeDocument[] starTreeDocuments) {
+        return new Iterator<>() {
+            boolean hasNext = true;
+            StarTreeDocument currentStarTreeDocument = starTreeDocuments[0];
+            int docId = 1;
+
+            @Override
+            public boolean hasNext() {
+                return hasNext;
+            }
+
+            @Override
+            public StarTreeDocument next() {
+                // aggregate as we move on to the next doc
+                StarTreeDocument next = reduceSegmentStarTreeDocuments(null, currentStarTreeDocument);
+                while (docId < starTreeDocuments.length) {
+                    StarTreeDocument starTreeDocument = starTreeDocuments[docId++];
```

Comment:
I think it can be assigned, but I feel it would be more readable if the `currentStarTreeDocument` is only reassigned when dimensions are not equal. wdyt?
Changing to explicit ++.